src/NHibernate/Dialect/SQLiteDialect.cs

```diff
@@ -38,32 +37,46 @@ public SQLiteDialect()
 
 		protected virtual void RegisterColumnTypes()
 		{
+			// SQLite really has only five types, and a very lax typing system, see https://www.sqlite.org/datatype3.html
+			// Please do not map (again) fancy types that do not actually exist in SQLite, as this is kind of supported by
+			// SQLite but creates bugs in convert operations.
 			RegisterColumnType(DbType.Binary, "BLOB");
-			RegisterColumnType(DbType.Byte, "TINYINT");
-			RegisterColumnType(DbType.Int16, "SMALLINT");
-			RegisterColumnType(DbType.Int32, "INT");
-			RegisterColumnType(DbType.Int64, "BIGINT");
+			RegisterColumnType(DbType.Byte, "INTEGER");
+			RegisterColumnType(DbType.Int16, "INTEGER");
+			RegisterColumnType(DbType.Int32, "INTEGER");
+			RegisterColumnType(DbType.Int64, "INTEGER");
 			RegisterColumnType(DbType.SByte, "INTEGER");
 			RegisterColumnType(DbType.UInt16, "INTEGER");
 			RegisterColumnType(DbType.UInt32, "INTEGER");
 			RegisterColumnType(DbType.UInt64, "INTEGER");
+
+			// NUMERIC and REAL are almost the same, they are binary floating point numbers. There is only a slight difference
+			// for values without a floating part. They will be represented as integers with numeric, but still as floating
+			// values with real. The side-effect of this is numeric being able of storing exactly bigger integers than real.
 			RegisterColumnType(DbType.Currency, "NUMERIC");
 			RegisterColumnType(DbType.Decimal, "NUMERIC");
```

Comment:
In #2079 I [had to change](https://github.com/nhibernate/nhibernate-core/blob/d3557db6ff12539087f103d46872ca987c06a95b/src/NHibernate/Dialect/SQLiteDialect.cs#L53) `Currency` and `Decimal` to REAL in order to correct division operator from prevent returing an integer when two non decimal values are divided (e.g. 19/2 would return 9 if NUMERIC is used). Maybe we can fix the issue in this PR instead.

Reply:
This will result in a loss of precision for big integer numbers stored into a decimal, since with current mapping, they are handled as int by SQLite, while they will be handled a binary float with the `REAL` typing. So this would be a possible breaking change for some corner cases. But it seems better than having that divide bug.